File: packages/stim-proto/src/discovery.ts
import type { Address, EndpointId, NodeId, ProtocolVersion } from "./ids.js";

export interface EndpointDeclaration {
  endpoint_id: EndpointId;
  node_id: NodeId;
  display_label?: string;
  endpoint_kind?: string;
  supported_protocol_versions: ProtocolVersion[];
  supported_carriers: string[];
  content_capabilities: string[];
  security_capabilities: string[];
  declared_features: string[];
}

export interface DiscoveryRecord {
  node_id: NodeId;
  endpoint_declaration: EndpointDeclaration;
  carrier_kind: string;
  addresses: Address[];
  protocol_versions: ProtocolVersion[];
}

export interface DeliveryTarget {
  node_id: NodeId;
  carrier_kind: string;
  selected_address: Address;
  protocol_version: ProtocolVersion;
}

File: crates/stim-proto/src/discovery.rs
use serde::{Deserialize, Serialize};
use utoipa::ToSchema;

use crate::{Address, EndpointId, NodeId, ProtocolVersion};

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, ToSchema)]
pub struct EndpointDeclaration {
    pub endpoint_id: EndpointId,
    pub node_id: NodeId,
    pub display_label: Option<String>,
    pub endpoint_kind: Option<String>,
    pub supported_protocol_versions: Vec<ProtocolVersion>,
    pub supported_carriers: Vec<String>,
    pub content_capabilities: Vec<String>,
    pub security_capabilities: Vec<String>,
    pub declared_features: Vec<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, ToSchema)]
pub struct DiscoveryRecord {
    pub node_id: NodeId,
    pub endpoint_declaration: EndpointDeclaration,
    pub carrier_kind: String,
    pub addresses: Vec<Address>,
    pub protocol_versions: Vec<ProtocolVersion>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, ToSchema)]
pub struct DeliveryTarget {
    pub node_id: NodeId,
    pub carrier_kind: String,
    pub selected_address: Address,
    pub protocol_version: ProtocolVersion,
}
